src/Datadog.Trace.ClrProfiler.Managed/Emit/ModuleLookup.cs

```diff
@@ -0,0 +1,62 @@
+using System;
+using System.Collections.Concurrent;
+using System.Reflection;
+using System.Threading;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.ClrProfiler.Emit
+{
+    internal static class ModuleLookup
+    {
+        private static readonly ILog Log = LogProvider.GetLogger(typeof(ModuleLookup));
+
+        private static ManualResetEventSlim _populationResetEvent = new ManualResetEventSlim(initialState: true);
+        private static ConcurrentDictionary<Guid, Module> _modules = new ConcurrentDictionary<Guid, Module>();
+
+        public static Module Get(Guid moduleVersionId)
+        {
+            Module value = null;
+
+            // Block if a population event is happening
+            _populationResetEvent.Wait();
+
+            if (_modules.TryGetValue(moduleVersionId, out value))
+            {
+                return value;
+            }
+
+            // Block threads on this event
+            _populationResetEvent.Reset();
+
+            try
+            {
+                PopulateModules();
+            }
+            catch (Exception ex)
+            {
+                Log.Error("Error when populating modules.", ex);
+            }
+            finally
+            {
+                // Continue threads blocked on this event
+                _populationResetEvent.Set();
+            }
+
+            _modules.TryGetValue(moduleVersionId, out value);
```

Comment:
Can we add some logging in case we don't find the MVID? Maybe something like:

```csharp
if (!modules.TryGetValue(moduleVersionId, out value))
{
    Log.Error($"Unable to find moduleVersionId {moduleVersionId}");
}
```

Reply:
Or, perhaps a different error message since this will be showing up in customer logs. I'm not sure what we want to display here.

Reply:
I was thinking about introducing a log into the MethodBuilder actually...
Something along the lines of, whenever the Module is null, mention that we are resolving with the fallback mechanism. 
It might be good to differentiate between using the fallback because we didn't find the right method in the Module we had, and just not having a module at all.

Reply:
That way we don't log every single time the lookup for a Module happens, only on the first Method Build.

Reply:
I agree we should differentiate between not finding the Module and finding the Module but using the fallback mechanism.

And logging on the first Method Build is fine. If it's not loaded by the time we get to the method execution, I'm not sure it will ever be loaded.

Reply:
Yeah I personally hope we never see that log statement